bridge-fungible-token/src/utils.sw

```diff
@@ -1,11 +1,184 @@
+    mint_to_address(amount, to);
+    log(MintEvent { amount, to });
+}
+
+pub fn burn_tokens(amount: u64, from: Identity) {
+    burn(amount);
+    log(BurnEvent {
+        from: from,
+        amount,
+    })
+}
+
+pub fn correct_input_type(index: u64) -> bool {
+    let type = input_type(index);
+    match type {
+        Input::Message => {
+            true
+        },
+        _ => {
+            false
+        }
+    }
+}
+
+pub fn safe_b256_to_u64(val: b256) -> Result<u64, BridgeFungibleTokenError> {
+    // first, decompose into u64 values
+    let u64s = decompose(val);
+    log(u64s.3);
+
+    // verify amount will require no partial refund of dust by ensuring that
+    // the first 9 decimal places in the passed-in value are empty,
+    // then verify amount is not too small or too large
+    if (u64s.3 / 1_000_000_000) * 1_000_000_000 == u64s.3
+        && u64s.3 >= 1_000_000_000
+        && u64s.0 == 0
+        && u64s.1 == 0
+        && u64s.2 == 0
+    {
+        // reduce decimals by 9 places
+        Result::Ok(u64s.3 / 1_000_000_000)
+    } else {
+        Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+    }
+}
+
+/// Build a single b256 value from a u64 left-padded with 3 0u64's
+pub fn u64_to_b256(val: u64) -> b256 {
+    let res: b256 = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    let ptr = addr_of(res);
+    write(ptr, 0);
+    write(ptr + 8, 0);
+    write(ptr + 16, 0);
+    write(ptr + 24, val);
+    res
+}
+
+/// Get 4 64 bit words from a single b256 value.
+pub fn decompose(val: b256) -> (u64, u64, u64, u64) {
+    let w1 = single_word_from_b256(val, 0);
+    let w2 = single_word_from_b256(val, 8);
+    let w3 = single_word_from_b256(val, 16);
+    let w4 = single_word_from_b256(val, 24);
+    (w1, w2, w3, w4)
+}
+
+/// Extract a single 64 bit word from a b256 value using the specified offset.
+fn single_word_from_b256(val: b256, offset: u64) -> u64 {
+    let mut empty: u64 = 0;
+    asm(r1: val, offset: offset, r2, res: empty) {
+        add r2 r1 offset;
+        lw res r2 i0;
+        res: u64
+    }
+}
+
+pub fn parse_message_data(msg_idx: u8) -> MessageData {
+    let mut msg_data = MessageData {
+        fuel_token: ~ContractId::from(ZERO_B256),
+        l1_asset: ~EvmAddress::from(ZERO_B256),
+        from: ~EvmAddress::from(ZERO_B256),
+        to: ~Address::from(ZERO_B256),
+        amount: ZERO_B256,
+    };
+
+    // Parse the message data
+    msg_data.fuel_token = ~ContractId::from(input_message_data::<b256>(msg_idx, 0));
+    msg_data.l1_asset = ~EvmAddress::from(input_message_data::<b256>(msg_idx, 32));
+    msg_data.from = ~EvmAddress::from(input_message_data::<b256>(msg_idx, 32 + 32));
+    msg_data.to = ~Address::from(input_message_data::<b256>(msg_idx, 32 + 32 + 32));
+    msg_data.amount = input_message_data::<b256>(msg_idx, 32 + 32 + 32 + 32);
+
+    msg_data
+}
+pub fn encode_data(to: b256, amount: u64) -> Vec<u64> {
```

Comment:
should the predicate_root be encoded in message.data ?

Reply:
no, this is the message going to Ethereum which doesn't care about predicates. The Ethereum side will really only care who the sender was (what Fuel contract is talking to them)